hal/src/dmac/channel/mod.rs

```diff
@@ -332,23 +277,71 @@ impl<Id: ChId> Channel<Id, Busy> {
     /// and BUSYCH is set. To make sure the transfer is actually complete, the
     /// channel needs to be both NOT PENDING and NOT BUSY.
     #[inline]
-    pub(crate) fn xfer_complete(&self, dmac: &DMAC) -> bool {
-        let id = Id::U8;
-        dmac.busych.read().bits() & (1 << id) == 0 && dmac.pendch.read().bits() & (1 << id) == 0
+    pub(crate) fn xfer_complete(&self) -> bool {
+        !self.regs.busych.read_bit() && !self.regs.pendch.read_bit()
```

Comment:
(have only checked the SAMD21 docs, perhaps other chips are different)

I think this actually reports whether a beat transfer is pending or ongoing; it'll return `true` if there are more beats to transfer but the channel is waiting on a trigger, for instance.  In that case at least, what's actually wanted is `!self.regs.chctrla.read().enable().bit_is_set()`.